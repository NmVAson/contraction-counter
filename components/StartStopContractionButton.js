import React, {Component} from 'react';
import {StyleSheet, View, Button, ScrollView, Text, Alert} from 'react-native';
import moment from 'moment';

const numberOfMinutesBetweenContractions = 5.5;
const numberOfMinutesForAContraction = .75;

export default class StartStopContractionButton extends Component {
  constructor(props) {
    super(props);
    this.state = { 
      title: 'click to start contraction',
      isContracting: false,
      contractions: [],
      currentStartTime: null
    };
  }

  toggleContraction() {
    let isStartingANewContraction = !this.state.isContracting;
    let newTitle;

    if(isStartingANewContraction) {
      this.state.contractions.push({start: moment(), end: null});
      newTitle = 'click to end contraction';
    } else {
      let currentContraction = this.state.contractions.pop();
      currentContraction.end = moment();
      this.state.contractions.push(currentContraction);

      newTitle = 'click to start contraction';
    }

    this.setState((state, props) => ({
      title: newTitle,
      isContracting: !state.isContracting
    }));
  }

  createContractionComponent(text) {
    return <Text className='contraction' style={[this.props.styles.text, this.props.styles.h5]}>{text}</Text>;
  }

  renderContraction(c, i, contractions) {
    if(!c.end && i ==0) return;

    if(i == 0) {
      let duration = moment.duration(c.end.diff(c.start));
      let durationLabel = moment.utc(duration.as('milliseconds')).format("H[h] m[m] s[s]");
      
      return this.createContractionComponent(`Duration: ${durationLabel}`);
    } else {
      let frequency = moment.duration(c.start.diff(contractions[i-1].start));
      let frequencyLabel = moment.utc(frequency.as('milliseconds')).format("H[h] m[m] s[s]");
      
      if(!c.end) {
        return this.createContractionComponent(`Frequency: ${frequencyLabel}`);
      }

      let duration = moment.duration(c.end.diff(c.start));
      let durationLabel = moment.utc(duration.as('milliseconds')).format("H[h] m[m] s[s]");
      return this.createContractionComponent(`Frequency: ${frequencyLabel}, Duration: ${durationLabel}`);
    }
  }

  render() {
    return (<View style={this.props.styles.h75}>
      <Button 
        title={this.state.title}
        onPress={this.toggleContraction.bind(this)}/>
      <ScrollView 
        ref={ref => this.scrollView = ref}
        onContentSizeChange={(contentWidth, contentHeight) => {        
            this.scrollView.scrollToEnd({animated: true});
        }}
        style={this.props.styles.mt4}>
        {this.state.contractions.map(this.renderContraction.bind(this))}
      </ScrollView>
    </View>);
  }
}
