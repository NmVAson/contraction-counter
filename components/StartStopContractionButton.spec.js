import React from 'react';
import {View, Button, Text, Alert} from 'react-native';
import StartStopContractionButton from './StartStopContractionButton';
import { advanceBy, clear } from 'jest-date-mock';

import TestRenderer from 'react-test-renderer';

const oneMinuteInMs = 60000;
const fiveMinutesInMs = 300000;
const oneHourInMs = 3600000;
const fakeStyles = {
  h75: null,
  mt4: null,
  text: null,
  h5: null
};
let component;

beforeEach(() => {
  let testInstance = TestRenderer.create(<StartStopContractionButton styles={fakeStyles}/>);
  component = testInstance.root;

  jest.useFakeTimers();
});

afterEach(() => {
  clear();
  jest.restoreAllMocks();
  jest.runAllTimers();
});

it('matches previous snapshot', () => {
  let tree = TestRenderer.create(<StartStopContractionButton styles={fakeStyles}/>).toJSON();
  
  expect(tree).toMatchSnapshot();
});

it('changed button text on click', () => {
  let button = component.findByType(Button);

  expect(button.props.title).toBe('click to start contraction');

  startContraction();

  expect(button.props.title).toBe('click to end contraction');

  stopContraction();

  expect(button.props.title).toBe('click to start contraction');
});

it('tracks the duration of the contraction', () => {
  startContraction();
  advanceBy(oneMinuteInMs);
  stopContraction();

  let durationTextList = findAllContractionComponents();
  expect(durationTextList).toHaveLength(1);
  let durationText = getText(durationTextList[0]);
  expect(durationText).toBe('Duration: 0h 1m 0s');
});

it('doesn\'t display duration if contraction hasn\'t ended', () => {
  startContraction();

  let durationTextList = findAllContractionComponents();
  expect(durationTextList).toHaveLength(0);
});

it('tracks the frequency of consecutive contractions', () => {
  startContraction();
  stopContraction();
  advanceBy(oneMinuteInMs);
  startContraction();

  let frequencyTextList = findAllContractionComponents();
  expect(frequencyTextList).toHaveLength(2);
  let frequencyText = getText(frequencyTextList[1]);
  expect(frequencyText).toBe('Frequency: 0h 1m 0s');
});

it('tracks the frequency and duration of complete contractions', () => {
  startContraction();
  stopContraction();
  advanceBy(oneMinuteInMs);
  startContraction();
  stopContraction();

  let frequencyTextList = findAllContractionComponents();
  expect(frequencyTextList).toHaveLength(2);
  let frequencyText = getText(frequencyTextList[1]);
  expect(frequencyText).toBe('Frequency: 0h 1m 0s, Duration: 0h 0m 0s');
});

it('should throw an alert when contractions get to be 5 minutes apart and 1 minute long', () => {
  let alertSpy = jest.spyOn(Alert, 'alert');

  startContraction();
  stopContraction();
  advanceBy(fiveMinutesInMs);
  startContraction();
  advanceBy(oneMinuteInMs);
  stopContraction();

  expect(alertSpy).toHaveBeenCalled();
  expect(alertSpy).toBeCalledWith('511! You\'ll want to head to the hospital in an hour!');
});

it('should throw an alert if it\'s been an hour since contractions are 5 minutes apart and 1 minute long', () => {
  let alertSpy = jest.spyOn(Alert, 'alert');
  let button = component.findByType(Button);
  startContraction();
  stopContraction();
  advanceBy(fiveMinutesInMs);
  startContraction();
  advanceBy(oneMinuteInMs);
  stopContraction();
  alertSpy.mockClear();

  jest.advanceTimersByTime(oneHourInMs - 1);

  expect(alertSpy).not.toHaveBeenCalled();

  jest.advanceTimersByTime(1);

  expect(alertSpy).toHaveBeenCalled();
  expect(alertSpy).toBeCalledWith('511! It\'s been an hour! Head to the hospital!');
});

function startContraction() {
  component.findByType(Button).props.onPress();
}

function stopContraction() {
  component.findByType(Button).props.onPress();
}

function findAllContractionComponents() {
  return component.findAll(n => n.type == 'Text' && n.props.className == 'contraction');
}

function getText(textComponent) {
  return textComponent.props.children
}