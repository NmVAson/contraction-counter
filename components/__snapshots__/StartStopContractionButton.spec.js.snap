// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches previous snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#9EC4C5",
      "borderRadius": 5,
      "opacity": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "padding": 10,
        },
        Object {
          "color": "#ffffff",
        },
        Object {
          "fontSize": 20,
          "padding": 20,
          "textAlign": "center",
        },
      ]
    }
  >
    click to start contraction
  </Text>
</View>
`;
